### DISCUSSION TITLE
Dynamic Routing

### DISCUSSION THREAD

**Discussion by mdmontesinos** [oqtane/oqtane.framework#2366] [Q&A] [answered]
I've been struggling to understand how to achievement dynamic routing in a similar sense that can be done with regular Blazor. For example, imagine I have a Razor component like this:

```c#
@page "\news\{Title}{Id:int}"

@((MarkupString)_markup)

@code {
[Parameter]
public string Title { get; set; }

[Parameter]
public int Id { get; set; }

private string _markup;

protected override async Task OnInitializedAsync()
    {
        await base.OnInitializedAsync();

        _markup = await CallNewsAPI(id); //it calls an external API that given the id, return the markup for the news
     }
}
```

I could dynamically render any news via the external api with just a single traditional razor component. 

In the case of Oqtane, I've thought of creating a module (similar to HtmlText) that could receive the Id, call the Api and render the markup, but I would still need to manually add and configure every single module. In addition, it is imperative that each news have its own routable page, and I don't quite understand how to do it with Oqtane custom routing, without having to manually add each page through Page Management .

**Comment by sbwalker** [comment, marked as answer]
Oqtane does not support all of the route template features that the default Blazor router supports. This is because Oqtane is multi-tenant and also because it retrieves its routes dynamically from the database (ie. so they are not hard-coded in the app code).

In regards to your example: 

@page "\news\\{Title}{Id:int}"

I am assuming "news" is the name of a page in your site/app. In Oqtane this would not be part of the route - it would be a virtual "page" that you create for your site. This abstracts the concept of a page away from the code itself and allows a variety of configuration to be associated to the page and managed at a higher level. For example at runtime the name of the page, the theme/layout, the permissions of who can view/edit, etc... can all be modified - allowing changes to occur without modifying code and deploying a new version of the application. There are many types of page related metadata that can be managed in this way. 

So this mean your actual static route is {Title}{Id:int} and I am assuming {Title} is the title of a specific news item and {Id:int} is its identifier. So basically this is a component which displays a specific news item based on the route information provided.

Oqtane has the concept of Url Parameters - which are different from QueryString parameters. The router parses these parameters out of the Url for you so that you can use them in your code. This is based on a delimiter character of "!". So a Url might look like:

/news/!/the-title-of-your-news-item/1

and to retrieve the Id:

var Id = int.Parse(PageState.UrlParameters.Split('/')[1])
or
var Id = int.Parse(GetUrlParameters("{Title}/{Id}")["Id"]);

and to construct a Url:

var url = NavigateUrl($"{PageState.Page.Path}/!/{encodedTitle}/{itemId}");

Url Parameters were introduced in June 2020 but are not being used extensively - in most cases QueryString parameters are being used instead. However if you want to create more friendly Urls this option is available. And looking at the code examples above, I can now imagine some simple enhancements to abstract the internals and make Url Parameters easier to use: 

var Id = int.Parse(PageState.UrlParameter[1]);
var url = NavigateUrl(AddUrlParameters(encodedTitle, itemId));

**Reply by sbwalker** [reply]
Added the AddUrlParameters() method in #2370

**Reply by mdmontesinos** [reply]
First of all, thanks for the great insights and examples about routing in Oqtane! I'll try to get the component working.

Secondly, I assume there's no way to get rid of the delimiter character ('!'), right? 

I'm asking because we currently need to redesign an active web application, where they use the routing I mentioned before, "news/{Title}/{Id}", but with Oqtane it would transform into "news/!/{Title}/{Id}", leading to possible mistakes for those who already access the site. In addition, as these routes are dynamic based on the id of the news (possible there are already hundreds of them), there's no way to create Url Mappings for all, unless it can be done dynamically as well.

**Reply by sbwalker** [reply]
You are correct there is no way to avoid the delimiter (!) when using Url Paramaters (see #638 for a long discussion about the feature). Oqtane does include a built in Url Mapping capability which would allow your current Urls to be mapped to new Urls. If a mapping is missed it will even log it as a Broken Link so that you are aware of the issue and can add the mapping. I had to deal with a lot of these when I migrated the oqtane.org site from DNN to Oqtane:

![image](https://user-images.githubusercontent.com/4840590/185115371-fa4244cc-c2d6-4829-90b2-12c7ec81e7f2.png)

(you will also discover a lot of crawlers coming to your site using strange Url patterns)

**Reply by mdmontesinos** [reply]
I guess the only viable option then is to create a Sql script that can automate the insertion of hundreds of Url Mappings into the DB.

Nonetheless, I consider this specific format of Url a major drawback (really needed to take into consideration) for anyone trying to migrate active sites made with old frameworks to Oqtane.

And as always, thanks for your time!

**Reply by sbwalker** [reply]
Url mapping is a required aspect of every web migration project I have ever worked on - usually migrating CMS content to another CMS. I can imagine the Url Mapping feature in Oqtane being enhanced in the future to accommodate pattern matching (ie. /news/ could be mapped to /news/!/ ).

**Reply by mdmontesinos** [reply]
That would be an excelent feature to have, extremely useful for complex sites.

**Reply by sbwalker** [reply]
One other item I failed to mention in my explanation yesterday is that routes are more complex in Oqtane than standard Blazor because Oqtane follows a module-centric model rather than a page-centric model. To put this into perspective, if you have a "home" page in Oqtane it might have numerous independent modules (ie. micro-frontends) which are rendered at runtime to create a composite "page". Each of these independent modules have their own functionality and may even have their own unique route requirements. So as a result, route management needs to occur at the module level rather than the page level.

**Reply by mdmontesinos** [reply]
That's right, I hadn't really considered how to both use "page" and "module" routing. I have yet to go deeper into the capabilities of the framework.

**Comment by mdmontesinos** [comment]
Why was the AddUrlParameters method changed? I specifically refer to this change #2372 :

![image](https://user-images.githubusercontent.com/90258222/188177379-efe365ff-751e-4769-bd1d-ed2031190454.png)

Before the change, I was using the auxiliar method like this:
```c#
var url = NavigateUrl(AddUrlParameters(encodedTitle, itemId));
```
And getting this url: localhost/news/!/title/id

But now the method just returns: localhost/title/id, which is basically incorrect and gets a 404.

I believe the method now is not quite as useful, as I would have to manually add the PageState.Page.Path and the Constants.UrlParameterDelimiter. It was much better before.

**Reply by sbwalker** [reply]
The AddUrlParameters() is a new method - it has never existed in any official Oqtane release. It was introduced for the upcoming 3.2 release but was modified based on further testing with additional scenarios. An example where it is used in the core framework is in Oqtane.Client\Modules\Admin\Logs\Index.razor:

```
NavigateUrl(PageState.Page.Path, AddUrlParameters(_level, _function, _rows, _page))
```

The string array was changed to an object array so that you do not need to cast every Url parameter (in the example above _rows and _page are integers). The usage was changed because NavigateUrl() already had legacy support for Url Parameters - as long as you passed them to one of the overloads which accept "parameters" as a parameter ie. NavigateUrl(string path, string parameters). So rather than creating a confusing API where in one instance you pass parameters as a path and another instance where you pass them as parameters, I felt it was better to be consistent. 

So for your example it would now be:

```
var url = NavigateUrl(PageState.Page.Path, AddUrlParameters(encodedTitle, itemId));
```

So you are correct that you now have to add the page path (but not the delimiter as that is abstracted within NavigateUrl already). 

I don't think its possible to include another overload of NavigateUrl(params object[] parameters) as the compiler would get confused.

**Reply by mdmontesinos** [reply]
RIght, thanks for the explanation! Sometimes it's tricky to figure out the appropriate methods to use in each case.

**Comment by sbwalker** [comment]
You might also be interested in how Url parameters are used in Oqtane.Client\Modules\Admin\Logs\Index.razor. New RouteTemplate and UrlParameters properties were added to ModuleBase, so in your code you define the structure of the Url Parameters and then you are able to simply reference the dictionary values:

```
	public override string RouteTemplate => "/{level}/{function}/{rows}/{page}";

...

			if (UrlParameters.ContainsKey("level"))
			{
				_level = UrlParameters["level"];
			}
```

**Reply by mdmontesinos** [reply]
Thanks for the advice! The RouteTemplate definitely is an interesting and useful concept, as it resembles (kind of) the usual routes that you would define in regular Blazor and obtain as Parameters, allowing for a more code-friendly dynamic routing.

**Reply by sbwalker** [reply]
I am still not sure if RouteTemplate is a good name or not... perhaps UrlParametersTemplate would be better so that it is clear that it only relates to Url Parameters and not any other route considerations. Thoughts?

**Reply by mdmontesinos** [reply]
That's probably a good change, as I was initially confused about what to actually include in the template. As it's named RouteTemplate, it seemed that perhaps I had to also include PageState.Path or something like that.

And by the way, I can't get the UrlParameters dictionary working.

I used the following RouteTemplate, just like Admin/Logs:
```c#
public override string RouteTemplate => "/{encodedTitle}/{id}";
```
However, UrlParameters.Keys.Count seems to be always 0, even though PageState.UrlParameters is not an empty string.

I used the following code:
```c#
Console.WriteLine("\n\nPageState.UrlParameters: " + PageState.UrlParameters);
Console.WriteLine("Url Parameters.KEYS COUNT: " + UrlParameters.Keys.Count + "\n\n");
```
And this was printed in the console:
```console
PageState.UrlParameters: mexico-top-medal-table-of-pkf-cadet--junior-and-u21-championships/2
Url Parameters.KEYS COUNT: 0
```
Therefore, UrlParameters are inserted into the Url, but the new Dictionary is not able to obtain them. Am I missing something?

**Reply by sbwalker** [reply]
I am guessing that it might depend on the life cycle event where your Console.Writeline() is implemented?

**Reply by mdmontesinos** [reply]
The code in inserted in the overrided method OnParametersSetAsync (but I've also tried using OnInitializedAsync). The loop for UrlParameters.Keys does never occur (because it's empty), and the conditional statement to check if the specific key is contained is always false. Consequently, the specific item that should be displayed based on the UrlParameters is always null.

```c#
protected override async Task OnParametersSetAsync()
    {
        try
        {
            Console.WriteLine("\n\nPageState.UrlParameters: " + PageState.UrlParameters);
            Console.WriteLine("Url Parameters.KEYS COUNT: " + UrlParameters.Keys.Count + "\n\n");

            foreach (var key in UrlParameters.Keys)
            {
                //Never enters here
                Console.WriteLine($"KEY: {key}, VALUE: {UrlParameters[key]}");
            }

            if (UrlParameters.ContainsKey("id") && int.TryParse(UrlParameters["id"], out int id))
            {
                //Never enters here
                Console.WriteLine("Found ID: " + id);
                _detailedNews = await WKFNewsService.GetWKFNewsAsync(id, ModuleState.ModuleId);
            }
            else
            {
                _WKFNewss = await WKFNewsService.GetWKFNewssAsync(ModuleState.ModuleId);
                _WKFNewss = _WKFNewss.OrderByDescending(x => x.CreatedOn).ToList();
            }
        }
        catch (Exception ex)
        {
            await logger.LogError(ex, "Error Loading WKFNews {Error}", ex.Message);
            ModuleInstance.AddModuleMessage(Localizer["Error Loading WKFNews"], MessageType.Error);
        }
    }
```

**Reply by sbwalker** [reply]
I will have to take a closer look... although it is strange that it is working fine in the Admin Logs area.

**Reply by sbwalker** [reply]
@mdmontesinos this should be fixed now in #2397. The new UrlParameters dictionary was using a local variable so that it did not need to recreate the dictionary everytime an item in the dictionary was referenced. The problem is that if the Url state was changed by the component it was not repopulating the dictionary. This would not manifest itself in the Admin Logs because the component does not change the Url state. Note that it important to use the OnParametersSetAsync() event when a component changes Url parameters as OnInitialized() is only called when a component is first instantiated. I also changed the name of the RouteTemplate to UrlParametersTemplate for API clarity as discussed earlier in this thread.

**Reply by mdmontesinos** [reply]
Great, it now works nicely! As always, awesome work.
